src/v/cluster/log_eviction_stm.cc

```diff
@@ -40,82 +41,114 @@ log_eviction_stm::log_eviction_stm(
-            /// There is still more data to truncate
+            co_await _queue.not_empty();
+        } catch (const ss::abort_requested_exception&) {
+            break;
         } catch (const ss::broken_condition_variable&) {
             /// stop() has been called
             break;
+        } catch (const ss::broken_semaphore&) {
+            /// stop() has been called
+            break;
         }
+
+        auto& event = _queue.front();
+        vlog(
+          _log.trace,
+          "processing prefix truncation event with offset: {}, waiting for "
+          "truncate: {}",
+          event.prefix_truncate_offset,
+          event.wait_for_truncation);
+
         auto evict_until = std::max(
-          _delete_records_eviction_offset, _storage_eviction_offset);
-        auto index_lb = _raft->log().index_lower_bound(evict_until);
-        if (!index_lb) {
-            wait_with_timeout = false;
+          _delete_records_eviction_offset, event.prefix_truncate_offset);
+        auto truncation_point = _raft->log().index_lower_bound(evict_until);
+        if (!truncation_point) {
+            if (!event.wait_for_truncation) {
+                maybe_pop_queue();
+            }
             continue;
         }
+
         vassert(
-          index_lb <= evict_until,
+          truncation_point <= evict_until,
           "Calculated boundary {} must be <= effective_start {} ",
-          index_lb,
+          truncation_point,
           evict_until);
         try {
-            co_await do_write_raft_snapshot(*index_lb);
+            co_await do_write_raft_snapshot(*truncation_point);
+            if (
+              !event.wait_for_truncation
+              || _raft->last_snapshot_index() == truncation_point) {
+                maybe_pop_queue();
+            } else {
+                vlog(
+                  _log.trace,
+                  "haven't yet been able to truncate at {} offset, waiting for "
+                  "next retry",
+                  event.prefix_truncate_offset);
+                co_await ss::sleep_abortable(5s, _as);
+            }
         } catch (const ss::abort_requested_exception&) {
             // ignore abort requested exception, shutting down
         } catch (const ss::gate_closed_exception&) {
             // ignore gate closed exception, shutting down
         } catch (const ss::broken_semaphore&) {
+        } catch (const ss::sleep_aborted&) {
             // ignore broken sem exception, shutting down
         } catch (const std::exception& e) {
             vlog(
-              _logger.error,
-              "Error occurred when attempting to write snapshot: "
-              "{}, ntp: {}",
-              e,
-              _raft->ntp());
+              _log.error,
+              "Error occurred when attempting to write snapshot: {}",
+              e);
         }
-        wait_with_timeout = _raft->last_snapshot_index() < index_lb;
     }
 }
 
+void log_eviction_stm::maybe_pop_queue() {
```

Comment:
how could `_queue` be empty when `pop` is ever called?

Reply:
the `abort` empties it out

Reply:
ah ok, then we just need to also catch exceptions that could arise from `push`